xds/src/main/java/io/grpc/xds/XdsRoutingLoadBalancer.java

```diff
@@ -69,54 +79,47 @@ public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
     logger.log(XdsLogLevel.DEBUG, "Received resolution result: {0}", resolvedAddresses);
     XdsRoutingConfig xdsRoutingConfig =
         (XdsRoutingConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
-    checkNotNull(xdsRoutingConfig, "Missing xds_routing lb config");
     Map<String, PolicySelection> newActions = xdsRoutingConfig.actions;
     for (String actionName : newActions.keySet()) {
       PolicySelection action = newActions.get(actionName);
-      if (!actions.containsKey(actionName)) {
-        RouteHelper routeHelper = new RouteHelper();
-        GracefulSwitchLoadBalancer routeBalancer = new GracefulSwitchLoadBalancer(routeHelper);
-        routeBalancer.switchTo(action.getProvider());
-        routeHelpers.put(actionName, routeHelper);
-        routeBalancers.put(actionName, routeBalancer);
-      } else if (!action.getProvider().equals(actions.get(actionName).getProvider())) {
-        routeBalancers.get(actionName).switchTo(action.getProvider());
+      if (!childLbStates.containsKey(actionName)) {
+        childLbStates.put(actionName, new ChildLbState(actionName, action.getProvider()));
+      } else {
+        childLbStates.get(actionName).reactivate(action.getProvider());
       }
+      childLbStates.get(actionName).lb
+          .handleResolvedAddresses(
```

Comment:
I would not execute `childLb.handleResolvedAddresses()` when `this.childLbStates` and `this.routes` are in a half-baked state. In extreme case `childLb.handleResolvedAddresses()` may trigger balancing state update inline.

Reply:
You are right. Fixed.